charts/openshift-metering: Update the reporting-operator and Presto StatefulSets to use TLS when specified

timflannagan1 opened the pull request:
### Overview:
Currently, the presto coordinator and worker pods communicate using http. With this change, the presto components can communicate using https when specified in the configurations, or by default, http.

You would need to specify key/cert pem files, which are then used to create a java keystore file and provided to presto.

chancez commented on charts/openshift-metering/templates/presto/presto-worker-config.yaml:
```diff
@@ -13,12 +13,29 @@ data:
+    node-scheduler.include-coordinator={{ .Values.presto.spec.presto.config.nodeSchedulerIncludeCoordinator }}
+    discovery-server.enabled=true
+
+    {{- if .Values.presto.spec.presto.config.useTLS }}
```
Indent is off

timflannagan1 replied:
It also references an invalid field. I updated both in the most recent commit.

chancez commented on charts/openshift-metering/templates/presto/presto-worker-config.yaml:
```diff
@@ -13,12 +13,29 @@ data:
+    http-server.https.port=8080
+    http-server.https.keystore.path=/opt/presto/tls/keystore.jks
+    http-server.https.keystore.key=password
+    node.internal-address=presto
```
This isn't correct, each worker will have a different address, this is the service name.

chancez replied:
Unfortunately, A records are not created for pod hostnames though, so we may need to create a headless service to make this work,

chancez replied:
https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-hostname-and-subdomain-fields has details on this. I think we want to set subdomain + create headless service so we can get an A record per presto pod. Then we can set `node.internal-address=FQDN`

chancez replied:
Update: we've decided to go with hostnames using statefulsets. So this can use `FQDN` instead.

chancez commented on charts/openshift-metering/templates/presto/presto-coordinator-config.yaml:
```diff
@@ -13,13 +13,29 @@ data:
+    http-server.https.port=8080
+    http-server.https.keystore.path=/opt/presto/tls/keystore.jks
+    http-server.https.keystore.key=password
+    node.internal-address=presto
```
I think we want to do this the same way that we do the workers, using the podIP. It shouldn't need to match the discovery URI.

chancez replied:
Update: we've decided to go with hostnames using statefulsets. So this can use `FQDN` instead.

timflannagan1 commented:
I added a commit to rename the Presto StatefulSet files after we merged the commit that changed them from a Deployment to a StatefulSet, and updated the `images/helm-operator/deploy-resources.sh` file to reflect that change. 

I removed the WIP tag, as the tests had previously passed, and this PR feels like it's in a good place now.

timflannagan1 commented:
/retest

chancez commented on charts/openshift-metering/templates/presto/presto-worker-config.yaml:
```diff
@@ -13,12 +13,28 @@ data:
+    internal-communication.https.keystore.key=password
+{{- else }}
     http-server.http.port=8080
+    discovery.uri={{ .Values.presto.spec.presto.config.discoveryURI }}
```
Honestly this could probably be hardcoded and removed from `values.yaml`

chancez commented on charts/openshift-metering/templates/presto/presto-coordinator-config.yaml:
```diff
@@ -13,13 +13,29 @@ data:
-    jmx.rmiserver.port=8081
-    jmx.rmiregistry.port=8081
     coordinator=true
+    node-scheduler.include-coordinator={{ .Values.presto.spec.presto.config.nodeSchedulerIncludeCoordinator }}
```
Honestly this could probably be hardcoded and removed from `values.yaml`

chancez replied:
Shoot my bad. I meant to put this on `discovery.uri`, sorry.

timflannagan1 replied:
Updated in the last force-push

chancez commented on charts/openshift-metering/templates/presto/presto-coordinator-statefulset.yaml:
```diff
@@ -52,10 +52,21 @@ spec:
+          value: /opt/presto/tls/keystore.pkcs12
+        - name: PRESTO_KEYSTORE_JKS
+          value: /opt/presto/tls/keystore.jks
+        - name: PASSWORD
```
`PRESTO_KEYSTORE_PASSWORD`

chancez commented:
This looks great. Can you write some brief instructions so I could test it out locally? Basically just the commands you would use to create the certs, secrets, and then what your custom Metering CR looks like.

chancez commented on charts/openshift-metering/values.yaml:
```diff
@@ -418,14 +422,15 @@ presto:
         hiveMetastoreURI: thrift://hive-metastore:9083
         maxQueryLength: "10000000"
         metastoreTimeout: null
-        nodeSchedulerIncludeCoordinator: true
```
We still want this

chancez commented on charts/openshift-metering/templates/reporting-operator/reporting-operator-config.yaml:
```diff
@@ -15,6 +15,9 @@ data:
   disable-prometheus-metrics-importer: {{ $operatorValues.spec.config.disablePrometheusMetricsImporter | quote}}
   enable-finalizers: {{ $operatorValues.spec.config.enableFinalizers | quote}}
+{{- if $operatorValues.spec.config.prestoTLS.enabled }}
+  presto-ca-file: "{{ $operatorValues.spec.config.prestoTLS.mountLocation }}/certificate.crt"
```
I feel like we should use ca.crt or tls.crt here for consistency. How about we just use `tls.crt` for now. 

Also, I'm not entirely sure why we have `$operatorValues.spec.config.prestoTLS.mountLocation` at all? I don't think any of this needs to be configurable. Additionally, maybe don't use `/var/run/secrets/java.io/keystores`, that's a bit confusing since we're not using a java keystore file in this secret, just a CA certificate and not a java program. Maybe use `/var/run/secrets/presto-tls/tls.crt`

chancez commented on charts/openshift-metering/templates/reporting-operator/reporting-operator-deployment.yaml:
```diff
@@ -257,6 +269,10 @@ spec:
 {{- end }}
         volumeMounts:
+{{- if $operatorValues.spec.config.prestoTLS.enabled }}
+        - name: presto-keystore-volume
```
`name: presto-tls` is probably more concise

timflannagan1 commented:
/retest

chancez commented on charts/openshift-metering/templates/presto/presto-worker-statefulset.yaml:
```diff
@@ -96,6 +115,10 @@ spec:
         - name: presto-logs
           mountPath: /var/presto/logs
+{{- if .Values.presto.spec.presto.config.tls.enabled }}
+        - name: presto-tls-volume
```
I don't think you need to put `-volume` on the end here, it's kinda implied.

chancez commented:
@timflannagan1  Can you squash this up a bit?

timflannagan1 commented:
@chancez rebased

chancez commented:
/lgtm
/approve

timflannagan1 commented:
/retest

chancez commented:
/retest

chancez commented:
/hold

chancez commented:
/hold cancel

chancez commented:
/lgtm
